src/compiler/types/update-import-refs.ts

```diff
@@ -4,75 +4,104 @@ import { dirname, resolve } from 'path';
 /**
  * Find all referenced types by a component and add them to the `importDataObj` parameter
  * @param importDataObj key/value of type import file, each value is an array of imported types
```

Comment:
Just reading through it looks like we're going to modify the `importDataObj` object that's passed in and then eventually return the same object. Would it make sense to clarify here in the JSDoc that that's what's going on? Just thinking that if I see a type signature like `func: (x: Foo, y: Bar) => Foo` I would tend to assume that the return `Foo` is going to be a new object (i.e. if I `let foo: Foo = { ... }; let otherFoo = func(foo);` then `foo === bar` should be `false`) but we might want to clarify that the return value is going to be the same object with new things added to it.

I guess this is a roundabout way of saying that I'm not sure what our attitude towards immutability is in Stencil - I generally find it is helpful to have immutable contracts between modules or other larger-scale 'interchanges' in the code, since then I don't have to know the implementation of a function defined in module `B` that I'm calling in module `A` to know that it's not going to mutate things that I pass in in an unexpected way, but at the same time often some well-scoped shared, mutable state is the simplest way to accomplish something in JS so it's not necessarily helpful to have a dogma of 'no mutation ever'.

Reply:
👍 updated in e2ebcf32c

As far as immutability goes...I don't think we can derive a hard and fast rule to uniformly apply to everything.  In general I agree with your point above about it not being helpful to be overly dogmatic here.  The problem that I think we will need to be aware of is that immutability can lead to a larger memory footprint.  When we handle data structures of certain sizes, that has the capability to crash the compiler (or more specifcially, the Node runtime runs out of memory, and _that_ crashes the compiler).  I don't think that's going to be a huge problem here, but have seen it lead to issues elsewhere in the codebase

Reply:
that makes sense, and going the immutable route does generally lead to a good deal of overhead w/ things like immer and whatnot. I'm for what gives us good performance!